### PR TITLE
feat(module): ensure there are no external ips in the cluster nodes

### PR DESCRIPTION
**Pull request by caetano-colin**
* This PR adds flags and tests to ensure nodes are using private ip addresses

### CODE REVIEW

**Comment by apeabody** [comment]
Hey @caetano-colin! - You will also need to include `master_authorized_networks`, I'm wondering if it will accept an empty list?  `master_authorized_networks = []`?

**Comment by caetano-colin** [comment]
> Hey @caetano-colin! - You will also need to include `master_authorized_networks`, I'm wondering if it will accept an empty list? `master_authorized_networks = []`? Otherwise we'll need to select a network.

The default value is an empty list, this are the transformations the module does with the variable

```terraform
master_authorized_networks_config = length(var.master_authorized_networks) == 0 ? [] : [{
    cidr_blocks : var.master_authorized_networks
  }]
 ```

https://github.com/terraform-google-modules/terraform-google-kubernetes-engine/blob/master/modules/beta-private-cluster/cluster.tf#L213

If think that because it is using a terraform dynamic block with an empty list it is not specifying an empty `master_authorized_networks_config` block like, but not specifying this argument at all:

```terraform
master_authorized_networks_config = {}
```

**Comment by caetano-colin** [comment]
as of commit 6e4b73dc272ce7309f2d4af2441f4ea214d02ab6 we removed the private cluster endpoint flag and tests

This means this PR now only contains tests/flags for ensuring private cluster nodes (i.e. nodes without external ip's)

File: 2-multitenant/modules/env_baseline/main.tf [hunk 1]
```diff
@@ -211,5 +214,8 @@ module "gke-autopilot" {
     module.eab_cluster_project
   ]
 
+  // Private Cluster Configuration
```
Comment (apeabody) [inline 2-multitenant/modules/env_baseline/main.tf:217]:
```suggestion
  // Private Cluster Configuration
```

**Comment by apeabody** [comment]
```
        	Error:      	Received unexpected error:
        	            	FatalError{Underlying: error while running command: exit status 1; 
        	            	Error: Error waiting for creating GKE cluster: Conflicting IP cidr range: Invalid IPCidrRange: 10.0.0.0/28 conflicts with existing subnetwork 'gke-cluster-us-central1-production-35d4d8fa-pe-subnet' in region 'us-central1'.
        	            	
        	            	  with module.env.module.gke-standard["eab-production-region02"].google_container_cluster.primary,
        	            	  on .terraform/modules/env.gke-standard/modules/beta-private-cluster/cluster.tf line 22, in resource "google_container_cluster" "primary":
        	            	  22: resource "google_container_cluster" "primary" {
        	            	}
        	Test:       	TestMultitenant/production
```

**Comment by caetano-colin** [comment]
> ```
>         	Error:      	Received unexpected error:
>         	            	FatalError{Underlying: error while running command: exit status 1; 
>         	            	Error: Error waiting for creating GKE cluster: Conflicting IP cidr range: Invalid IPCidrRange: 10.0.0.0/28 conflicts with existing subnetwork 'gke-cluster-us-central1-production-35d4d8fa-pe-subnet' in region 'us-central1'.
>         	            	
>         	            	  with module.env.module.gke-standard["eab-production-region02"].google_container_cluster.primary,
>         	            	  on .terraform/modules/env.gke-standard/modules/beta-private-cluster/cluster.tf line 22, in resource "google_container_cluster" "primary":
>         	            	  22: resource "google_container_cluster" "primary" {
>         	            	}
>         	Test:       	TestMultitenant/production
> ```

for the record, this is happening because there are more than one cluster being created and each of them is being created with a hardcoded CIDR (https://github.com/terraform-google-modules/terraform-google-kubernetes-engine/blob/master/modules/beta-private-cluster/variables.tf#L472), so when any other cluster is created after the first, we get an overlap

seeing if commit [d03efa4](https://github.com/GoogleCloudPlatform/terraform-google-enterprise-application/pull/212/commits/d03efa4e09cc003be37a10c5b37c245d0d50acdf) will fix it

**Comment by apeabody** [comment]
> Node pools are being created in an error state, at first I thought it might have been a firewall issue, as it was on my local tests.
> 
> is it possible we get a `export TF_LOG=DEBUG` or output of logs on the project to get more insight in why the node pool is on the error state?

Hi @caetano-colin - I triggered a fresh run so hopefully I can check the actual project.

**Comment by apeabody** [comment]
> > Node pools are being created in an error state, at first I thought it might have been a firewall issue, as it was on my local tests.
> > is it possible we get a `export TF_LOG=DEBUG` or output of logs on the project to get more insight in why the node pool is on the error state?
> 
> Hi @caetano-colin - I triggered a fresh run so hopefully I can check the actual project.

Looking like a networking problem, the node VMs appear to be unreachable.

File: 2-multitenant/modules/env_baseline/firewall.tf [hunk 2]
```diff
@@ -0,0 +1,53 @@
+resource "google_compute_firewall" "allow_internal_ingress" {
```
Comment (apeabody) [inline 2-multitenant/modules/env_baseline/firewall.tf:1]: Hey @caetano-colin - If you are looking to apply these to the cluster nodes, you will something like `network_tags = [allow_internal_ingress]` added to the cluster definitions below.

That said, I might suggest adding `add_cluster_firewall_rules = true` for now to the cluster definitions.

File: 2-multitenant/modules/env_baseline/variables.tf [hunk 3]
```diff
@@ -40,6 +40,11 @@ variable "network_project_id" {
   type        = string
 }
 
+variable "network_self_link" {
```
Comment (apeabody) [inline 2-multitenant/modules/env_baseline/variables.tf:43]: Looks like these `network_self_link` can be reverted out

File: test/setup/main.tf [hunk 4]
```diff
@@ -183,6 +185,48 @@ module "vpc" {
   }
 }
 
+resource "google_compute_firewall" "allow_private_google_access" {
```
Comment (apeabody) [inline test/setup/main.tf:188]: If these are needed, can we instead use the `terraform-google-network` module above's `ingress_rules` and `egress_rules` variables?

Comment (caetano-colin) [reply test/setup/main.tf]: resolved on c637f8a6f8c5e6b70d0df621c2fcd8b7cb93ff47

File: 2-multitenant/modules/env_baseline/firewall.tf (hunk 2 above)
Comment (caetano-colin) [reply 2-multitenant/modules/env_baseline/firewall.tf]: outdated, seems like we don't actually need the firewall rules

File: 2-multitenant/envs/development/main.tf [hunk 5]
```diff
@@ -27,6 +27,7 @@ module "env" {
   folder_id           = var.envs[local.env].folder_id
   network_project_id  = var.envs[local.env].network_project_id
   billing_account     = var.envs[local.env].billing_account
+  network_self_link   = var.envs[local.env].network_self_link
```
Comment (apeabody) [inline 2-multitenant/envs/development/main.tf:30]: Probably want to remove

File: 2-multitenant/envs/nonproduction/main.tf [hunk 6]
```diff
@@ -28,4 +28,5 @@ module "env" {
   network_project_id  = var.envs[local.env].network_project_id
   billing_account     = var.envs[local.env].billing_account
   cluster_subnetworks = var.envs[local.env].subnets_self_links
+  network_self_link   = var.envs[local.env].network_self_link
```
Comment (apeabody) [inline 2-multitenant/envs/nonproduction/main.tf:31]: Probably want to remove

File: 2-multitenant/envs/production/main.tf [hunk 7]
```diff
@@ -28,4 +28,5 @@ module "env" {
   network_project_id  = var.envs[local.env].network_project_id
   billing_account     = var.envs[local.env].billing_account
   cluster_subnetworks = var.envs[local.env].subnets_self_links
+  network_self_link   = var.envs[local.env].network_self_link
```
Comment (apeabody) [inline 2-multitenant/envs/production/main.tf:31]: Probably want to remove

File: 2-multitenant/modules/env_baseline/README.md [hunk 8]
```diff
@@ -24,6 +24,7 @@ The following resources are created:
 | env | The environment to prepare (ex. development) | `string` | n/a | yes |
 | folder\_id | Folder ID | `string` | n/a | yes |
 | network\_project\_id | Network Project ID | `string` | n/a | yes |
+| network\_self\_link | Network Self Link | `string` | n/a | yes |
```
Comment (apeabody) [inline 2-multitenant/modules/env_baseline/README.md:27]: Probably want to remove

File: 2-multitenant/modules/env_baseline/main.tf [hunk 9]
```diff
@@ -19,6 +19,14 @@ locals {
   subnetworks_re     = "/subnetworks/([^/]*)$"
   projects_re        = "projects/([^/]*)/"
   cluster_project_id = data.google_project.eab_cluster_project.project_id
+  available_cidr_ranges = [
```
Comment (apeabody) [inline 2-multitenant/modules/env_baseline/main.tf:22]: @gtsorbo - Thoughts if the private node range should be defined in a different stage/location?  For example, test/setup and passed through?

File: test/setup/main.tf [hunk 10]
```diff
@@ -145,16 +145,50 @@ module "vpc" {
...
+      name     = "allow-private-google-access"
+      priority = 200
+      destination_ranges = [
+        "34.126.0.0/18",
+        "199.36.153.8/30",
+      ]
+      allow = [
+        {
+          protocol = "tcp"
+          ports    = ["443"]
```
Comment (apeabody) [inline test/setup/main.tf:159]: Is there public documentation for allowing just port 443?  I found docs for the range and TCP (HTTP, HTTPS, HTTPS/2), but not a specific (HTTPS) port number.

- https://cloud.google.com/vpc/docs/configure-private-google-access
- https://cloud.google.com/vpc-service-controls/docs/set-up-private-connectivity

**Review by apeabody — APPROVED** [review]
Thanks @caetano-colin!

This LGTM, added a few design questions for @gtsorbo's final approval.

**Comment by caetano-colin** [comment]
Thanks! I'll implement these improvements

File: test/setup/main.tf (hunk 10 above)
Comment (caetano-colin) [reply test/setup/main.tf:159]: There is a doc saying that all traffic to `restricted.googleapis.com` is to port 443: https://cloud.google.com/vpc-service-controls/docs/set-up-private-connectivity#configure-firewall (but I have not found the same information for `private.googleapis.com`)

In the `terraform-example-foundation` the firewall rule for the Private Google Access network is using only port 443 also: https://github.com/terraform-google-modules/terraform-example-foundation/blob/master/3-networks-dual-svpc/modules/base_shared_vpc/firewall.tf#L58

File: 2-multitenant/modules/env_baseline/main.tf (hunk 9 above)
Comment (caetano-colin) [reply 2-multitenant/modules/env_baseline/main.tf]: Commit: 
a5a6d4a externalizes the blocks to a variable, if we are to define this ranges in a different stage we can use the variable.

Comment (apeabody) [reply 2-multitenant/modules/env_baseline/main.tf]: Looks good - Thanks!

File: test/setup/main.tf (hunk 10 above)
Comment (apeabody) [reply test/setup/main.tf:159]: Thanks! I think this is sufficient for now.

**Review by gtsorbo — APPROVED** [review]

**Pull request merged**